PR for suggestions and review purposes (#1)
NLnetLabs/draft-homburg-dnsop-igadp: dns-auth-proxy.md

%%%
title = "Implementation Guidelines for Authoritative DNS Proxies"
abbrev = "igadp"
area = "Internet"
workgroup = "DNSOP"

[seriesInfo]
status = "standard"
name = "Internet-Draft"
value = "draft-homburg-dnsop-igadp-00"
stream = "IETF"

date = 2023-07-06T00:00:00Z

[[author]]
initials="P.C."
surname="Homburg"
fullname="Philip Homburg"
organisation = "NLnet Labs"
  [author.address]
  email = "philip@nlnetlabs.nl"
%%%

.# Abstract

> In some situations it can attractive to have an authoritative DNS server that 

Author: ```suggestion
In some situations it can be attractive to have an authoritative DNS server that 
```

does not have a local copy of the zone or zones that it serves.

> In particular in anycast operations, it is attractive to have a great 

Author: ```suggestion
In particular in anycast operations, it is sensible to have a great 
```
Prevent repetition of the word attractive

geographical and topological diversity. However, sometimes the expect use of

> a particular site does not warrant the cost of keep local copies of the

Author: ```suggestion
geographical and topological diversity. However, sometimes the expected use of
a particular site does not warrant the cost of keeping local copies of the
```

> zones up-to-date. 

Author: ```suggestion
zones. 
```

This can be the case if the zone is very big or if the anycast cluster serves

> many zones and only a few on them are expect to get significant use.

Author: ```suggestion
This can be the case if a zone is very large or if the anycast cluster serves
many zones from which only a few are expected to receive significant traffic.
```

> In these cases it is attractive to have a proxy serve some or all of the zones.

Author: ```suggestion
In these cases it can be useful to have a proxy serve some or all of the zones.
```

> The proxy does not have a local copy of the zones it serves, instead it

Author: ```suggestion
The proxy would not have a local copy of the zones it serves, instead it
```

> forwards request to another server tat is authoritaive for the zone. 

Author: ```suggestion
forwards request to another server that is authoritative for the zone. 
```

The proxy may have a cache.
This document describes the details of such proxies.

{mainmatter}

# Discussion Venues

This note is to be removed before publishing as an RFC.
Source for this draft and an issue tracker can be found at
https://github.com/NLnetLabs/draft-homburg-dnsop-igadp.git .

# Introduction

> In some situations it can attractive to have an authoritative DNS server that 

Author: ```suggestion
In some situations it can be attractive to have an authoritative DNS server that 
```

does not have a local copy of the zone or zones that it serves.

> In particular in anycast operations, it is attractive to have a great 

Author: ```suggestion
In particular in anycast operations, it is sensible to have a great 
```

geographical and topological diversity.

> Benefits of an extra site can be reduced latency, DDoS protection by 

Author: ```suggestion
Benefits of an extra site can be: reduced latency, DDoS protection by 
```

serving a zone locally and thus be protected from a DDoS that originates in
other parts of the network, and DDoS mitigation by attracting local DDoS
traffic.

However, sometimes the expect use of a particular site does not warrant the

> cost of keep local copies of the zones up-to-date. 

Author: ```suggestion
However, sometimes the expected use of a particular site does not warrant the
cost of keeping local copies of the zones. 
```

This can be the case if the zone is very big or if the anycast cluster serves

> many zones and only a few on them are expect to get significant use.

Author: ```suggestion
This can be the case if a zone is very large or if the anycast cluster serves
many zones and only a few on them are expected to receive significant traffic.
```

> In these cases it is attractive to have a proxy serve some or all of the zones.

Author: ```suggestion
In these cases it can be useful to have a proxy serve some or all of the zones.
```

The proxy does not have a local copy of the zones it serves, instead it

> forwards request to another server tat is authoritaive for the zone. 

Author: ```suggestion
Instead of having a local copy of the zones, such a proxy 
forwards request to another server that is authoritative for the zone. 
```

The proxy may have a cache.

A proxy can operate in two basic modes: with or without cache. Without a cache,
the proxy forward each incoming request to another that is authoritative 

> for te domain. This is very simple, but also provides limited benefits.

Author: ```suggestion
A proxy can operate with or without cache. Without a cache,
the proxy forwards each incoming request to a name server that is authoritative 
for the domain. This is straight forward, but provides limited benefits.
```

With a cache the proxy becomes more complex. The reason is that we
want the proxy to behave like an authoritative server, i.e. provide similar

> behavior with respect to TTLs and data consistency.

Author: ```suggestion
Because we need the proxy to behave like an authoritative server and provide similar
behavior with respect to TTLs and data consistency, proxy behavior *with a cache* is more involved.
```

> Another issue that may complicate matters is the EDNS Client Subnet option.

Author: ```suggestion
Another issue that may complicate caching of responses is the EDNS Client Subnet option.
```

# Basic Requirements

> The proxy MUST drop replies with the AA bit clear. The reasoning is that
> the proxy may accidentally reach a recursive resolver.
> 

Author: ```suggestion
```
Delegation responses have the AA bit clear

The proxy sends requests upstream with RD bit clear. 

The proxy sends replies to clients with the original TTL. Caching is not 
controlled by TTL but by the parameters in the SOA record.

# Cache replacement

> The goal of the cache replacement strategy described here is mimic the 

Author: ```suggestion
The goal of the cache replacement strategy described here is to mimic the 
```

behavior of a standard secondary as much as possible.

> The basic idea is that the proxy fetches SOA records just like a secondary
> would do based on the parameters in the SOA record, in response to a NOTIFY
> and taking the contents of the EXPIRE option RFC 7314.

Author: ```suggestion
The proxy fetches SOA records just like a secondary
would do based on the parameters in the SOA record in response to a NOTIFY,
and takes the contents of the EXPIRE option [@!RFC7314].
```

> Cache entries are tags with the zone serial number from the most recent SOA
> record. A cache entry is valid only when the serial number in the cache
> entry match the serial number of the current SOA record.

Author: ```suggestion
Cache entries are tagged with the zone serial number from the most recent SOA
record. A cache entry is valid only when the serial number in the cache
entry matches the serial number of the current SOA record.
```

> When a new SOA record arrives with a higher serial number there are two

Author: ```suggestion
When a new SOA record arrives with a higher serial number, there are two
```

options. 
The first option is to take this as the new current SOA record. This has the
effect of immediately invalidating the cache.

The second option is to delay accepting this SOA and load new copies of
hot cache items first.

> Note that DNS replies may contain SOA records. If a DNS replies contains a
> SOA record with a higher serial number then the proxy updates its copy of
> the SOA records using one of the two techniques descibed above.

Author: ```suggestion
Note that DNS replies may contain SOA records. If a DNS replies contains a
SOA record with a higher than the most recently seen serial number (taking into account Serial Number Arithmetic [@!RFC1982]), then the proxy updates its copy of
the SOA records using one of the two techniques described above.
```

Maybe also include something on serial arithmetics

# Aggressive negative caching

> A caching proxy is allowed to generate negative responses (NODATA or NXDOMAIN)
> based on information in NSEC or NSEC3 records. Obviously this can be done 

Author: ```suggestion
A caching proxy is allowed to generate negative responses (NODATA or NXDOMAIN)
[@!RFC8198;@RFC9077] based on information in NSEC or NSEC3 records. Obviously this can be done 
```

only for DNSSEC signed zones, and NXDOMAIN only for NSEC or NSEC3 without
opt-out.

# ENDS Client Subnet Option

> A proxy MAY support RFC 7871 (Client Subnet in DNS Queries). Support is optional because many authoritative server have no need to know the IP address of the client.

Author: ```suggestion
A proxy MAY support [@!RFC7871] (Client Subnet in DNS Queries).
```
